VK crashes when populating env var from non-existing secret

---

### Environment summary

Provider (triton)

Version (0)

K8s Master Info (minikube)

### Issue Details
If you provide a secretKeyRef that doesn't exist VK dumps.

```
/usr/local/go/src/runtime/asm_amd64.s:1333
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x108 pc=0x19d31c6]

goroutine 64 [running]:
github.com/virtual-kubelet/virtual-kubelet/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x108                                                                                                                       
panic(0x1cd5340, 0x3aa7af0)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/virtual-kubelet/virtual-kubelet/vkubelet.(*Server).populateEnvironmentVariables(0xc0003aa230, 0xc000cc5880, 0xc0002e0c60, 0x11)                                                                                                                   
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vkubelet/env.go:41 +0x216
github.com/virtual-kubelet/virtual-kubelet/vkubelet.(*Server).createOrUpdatePod(0xc0003aa230, 0x23ba960, 0xc000b05b30, 0xc000cc5880, 0x0, 0x0)                                                                                                               
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vkubelet/pod.go:43 +0x16e
github.com/virtual-kubelet/virtual-kubelet/vkubelet.(*PodController).syncPodInProvider(0xc000332be0, 0x23ba960, 0xc000af7f20, 0xc000cc5880, 0x0, 0x0)                                                                                                        
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vkubelet/podcontroller.go:286 +0x2c4
github.com/virtual-kubelet/virtual-kubelet/vkubelet.(*PodController).syncHandler(0xc000332be0, 0x23ba960, 0xc000af7ec0, 0xc000c30d20, 0x1f, 0x0, 0x0)                                                                                                        
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vkubelet/podcontroller.go:257 +0x748
github.com/virtual-kubelet/virtual-kubelet/vkubelet.(*PodController).processNextWorkItem.func1(0xc000332be0, 0x23ba960, 0xc000af7e60, 0xc000659800, 0x1beff80, 0xc000170580, 0x0, 0x0)                                                                       
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vkubelet/podcontroller.go:195 +0x1af
github.com/virtual-kubelet/virtual-kubelet/vkubelet.(*PodController).processNextWorkItem(0xc000332be0, 0x23ba960, 0xc000af7e60, 0x215a406, 0x2, 0x0)                                                                                                         
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vkubelet/podcontroller.go:209 +0x20b
github.com/virtual-kubelet/virtual-kubelet/vkubelet.(*PodController).runWorker(0xc000332be0, 0x23ba8a0, 0xc000a0c100, 0x215a406, 0x2)                                                                                                                        
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vkubelet/podcontroller.go:155 +0x53
github.com/virtual-kubelet/virtual-kubelet/vkubelet.(*PodController).Run.func1()
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vkubelet/podcontroller.go:142 +0x7e
github.com/virtual-kubelet/virtual-kubelet/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000a26de0)
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54                                                                                                                             
github.com/virtual-kubelet/virtual-kubelet/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000a26de0, 0x3b9aca00, 0x0, 0x1, 0xc000056540)                                                                                                            
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbe                                                                                                                             
github.com/virtual-kubelet/virtual-kubelet/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc000a26de0, 0x3b9aca00, 0xc000056540)                                                                                                                            
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d                                                                                                                              
created by github.com/virtual-kubelet/virtual-kubelet/vkubelet.(*PodController).Run
        /home/smith/go/src/github.com/virtual-kubelet/virtual-kubelet/vkubelet/podcontroller.go:140 +0x223
```


### Repo Steps


Configure a pod / replicaset / or deployment and provide a secretKeyRef that doesn't exist.

```
    spec:                                                                                                                                                                                                                                                     
      restartPolicy: Always                                                                                                                                                                                                                                   
      containers:                                                                                                                                                                                                                                             
      - image: c239a01f-5fe9-456f-8f92-53f5b7efd53d                                                                                                                                                                                                           
        name: consul                                                                                                                                                                                                                                          
        ports:                                                                                                                                                                                                                                                
          - containerPort: 8500                                                                                                                                                                                                                               
            protocol: TCP                                                                                                                                                                                                                                     
            name: consul                                                                                                                                                                                                                                      
          - containerPort: 22                                                                                                                                                                                                                                 
            name: ssh                                                                                                                                                                                                                                         
        envFrom:                                                                                                                                                                                                                                              
          - configMapRef:                                                                                                                                                                                                                                     
              name: consul-config                                                                                                                                                                                                                             
        env:                                                                                                                                                                                                                                                  
        - name: SECRET_USERNAME                                                                                                                                                                                                                               
          valueFrom:                                                                                                                                                                                                                                          
           secretKeyRef:                                                                                                                                                                                                                                      
              name: mysecrets                                                                                                                                                                                                                                 
              key: username     
```

--- a/test/e2e/basic_test.go
+++ b/test/e2e/basic_test.go
@@ -189,6 +189,128 @@ func TestPodLifecycle(t *testing.T) {
 	}
 }
 
+// TestCreatePodWithOptionalInexistantSecrets tries to create a pod referencing optional, inexistant secrets.
+// It then verifies that the pod is created successfully.
+func TestCreatePodWithOptionalInexistantSecrets(t *testing.T) {
+	// Create a pod with a single container referencing optional, inexistant secrets.
+	pod, err := f.CreatePod(f.CreatePodObjectWithOptionalSecretKey())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Delete the pod after the test finishes.
+	defer func() {
+		if err := f.DeletePod(pod.Namespace, pod.Name); err != nil && !apierrors.IsNotFound(err) {
+			t.Error(err)
+		}
+	}()
+
+	// Wait for the pod to be reported as running and ready.
+	if err := f.WaitUntilPodReady(pod.Namespace, pod.Name); err != nil {
+		t.Fatal(err)
+	}
+
+	// Check that the pod is known to the provider.
+	stats, err := f.GetStatsSummary()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := findPodInPodStats(stats, pod); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestCreatePodWithMandatoryInexistantSecrets tries to create a pod referencing inexistant secrets.
+// It then verifies that the pod is not created.
+func TestCreatePodWithMandatoryInexistantSecrets(t *testing.T) {
+	// Create a pod with a single container referencing inexistant secrets.
+	pod, err := f.CreatePod(f.CreatePodObjectWithMandatorySecretKey())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Delete the pod after the test finishes.
+	defer func() {
+		if err := f.DeletePodImmediately(pod.Namespace, pod.Name); err != nil && !apierrors.IsNotFound(err) {
+			t.Error(err)
+		}
+	}()
+
+	// Wait for a small period so we can later assert the pod is not running.
+	// TODO (@pires) Wait for event to be reported.
+	time.Sleep(5 * time.Second)
+
+	// Check that the pod is NOT known to the provider.
+	stats, err := f.GetStatsSummary()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := findPodInPodStats(stats, pod); err == nil {
+		t.Fatalf("Expecting to NOT find pod \"%s/%s\" having mandatory, inexistant secrets.", pod.Namespace, pod.Name)
+	}
+}
+
+// TestCreatePodWithOptionalInexistantConfigMap tries to create a pod referencing optional, inexistant config map.
+// It then verifies that the pod is created successfully.
+func TestCreatePodWithOptionalInexistantConfigMap(t *testing.T) {
+	// Create a pod with a single container referencing optional, inexistant config map.
+	pod, err := f.CreatePod(f.CreatePodObjectWithOptionalConfigMapKey())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Delete the pod after the test finishes.
+	defer func() {
+		if err := f.DeletePod(pod.Namespace, pod.Name); err != nil && !apierrors.IsNotFound(err) {
+			t.Error(err)
+		}
+	}()
+
+	// Wait for the pod to be reported as running and ready.
+	if err := f.WaitUntilPodReady(pod.Namespace, pod.Name); err != nil {
+		t.Fatal(err)
+	}
+
+	// Check that the pod is known to the provider.
+	stats, err := f.GetStatsSummary()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := findPodInPodStats(stats, pod); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestCreatePodWithMandatoryInexistantConfigMap tries to create a pod referencing inexistant secrets.
+// It then verifies that the pod is not created.
+func TestCreatePodWithMandatoryInexistantConfigMap(t *testing.T) {
+	// Create a pod with a single container referencing inexistant config map.
+	pod, err := f.CreatePod(f.CreatePodObjectWithMandatoryConfigMapKey())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Delete the pod after the test finishes.
+	defer func() {
+		if err := f.DeletePodImmediately(pod.Namespace, pod.Name); err != nil && !apierrors.IsNotFound(err) {
+			t.Error(err)
+		}
+	}()
+
+	// Wait for a small period so we can later assert the pod is not running.
+	// TODO (@pires) Wait for event to be reported.
+	time.Sleep(5 * time.Second)
+
+	// Check that the pod is NOT known to the provider.
+	stats, err := f.GetStatsSummary()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := findPodInPodStats(stats, pod); err == nil {
+		t.Fatalf("Expecting to NOT find pod \"%s/%s\" having mandatory, inexistant config map.", pod.Namespace, pod.Name)
+	}
+}
+
 // findPodInPodStats returns the index of the specified pod in the .pods field of the specified Summary object.
 // It returns an error if the specified pod is not found.
 func findPodInPodStats(summary *v1alpha1.Summary, pod *v1.Pod) (int, error) {

--- a/test/e2e/framework/env.go
+++ b/test/e2e/framework/env.go
@@ -1,0 +1,81 @@
+package framework
+
+import (
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	bFalse = false
+	bTrue  = true
+)
+
+// CreatePodObjectWithMandatoryConfigMapKey creates a pod object that references the "key_0" key from the "config-map-0" config map as mandatory.
+func (f *Framework) CreatePodObjectWithMandatoryConfigMapKey() *corev1.Pod {
+	return f.CreatePodObjectWithEnv([]corev1.EnvVar{
+		{
+			Name: "CONFIG_MAP_0_KEY_0",
+			ValueFrom: &corev1.EnvVarSource{
+				ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{Name: "config-map-0"},
+					Key:                  "key_0",
+					Optional:             &bFalse,
+				},
+			},
+		},
+	})
+}
+
+// CreatePodObjectWithOptionalConfigMapKey creates a pod object that references the "key_0" key from the "config-map-0" config map as optional.
+func (f *Framework) CreatePodObjectWithOptionalConfigMapKey() *corev1.Pod {
+	return f.CreatePodObjectWithEnv([]corev1.EnvVar{
+		{
+			Name: "CONFIG_MAP_0_KEY_0",
+			ValueFrom: &corev1.EnvVarSource{
+				ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{Name: "config-map-0"},
+					Key:                  "key_0",
+					Optional:             &bTrue,
+				},
+			},
+		},
+	})
+}
+
+// CreatePodObjectWithMandatorySecretKey creates a pod object that references the "key_0" key from the "secret-0" config map as mandatory.
+func (f *Framework) CreatePodObjectWithMandatorySecretKey() *corev1.Pod {
+	return f.CreatePodObjectWithEnv([]corev1.EnvVar{
+		{
+			Name: "SECRET_0_KEY_0",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{Name: "secret-0"},
+					Key:                  "key_0",
+					Optional:             &bFalse,
+				},
+			},
+		},
+	})
+}
+
+// CreatePodObjectWithOptionalSecretKey creates a pod object that references the "key_0" key from the "secret-0" config map as optional.
+func (f *Framework) CreatePodObjectWithOptionalSecretKey() *corev1.Pod {
+	return f.CreatePodObjectWithEnv([]corev1.EnvVar{
+		{
+			Name: "SECRET_0_KEY_0",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{Name: "secret-0"},
+					Key:                  "key_0",
+					Optional:             &bTrue,
+				},
+			},
+		},
+	})
+}
+
+// CreatePodObjectWithEnv creates a pod object whose name starts with "env-test-" and that uses the specified environment configuration for its first container.
+func (f *Framework) CreatePodObjectWithEnv(env []corev1.EnvVar) *corev1.Pod {
+	pod := f.CreateDummyPodObjectWithPrefix("env-test-", "foo")
+	pod.Spec.Containers[0].Env = env
+	return pod
+}

--- a/vkubelet/env.go
+++ b/vkubelet/env.go
@@ -2,7 +2,6 @@ package vkubelet
 
 import (
 	"fmt"
-
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 )
@@ -15,34 +14,87 @@ func (s *Server) populateEnvironmentVariables(pod *corev1.Pod) error {
 				// Populate ConfigMaps to Env
 				if e.ValueFrom.ConfigMapKeyRef != nil {
 					vf := e.ValueFrom.ConfigMapKeyRef
+					// Check whether the key reference is optional.
+					// This will control whether we fail when unable to read the requested key.
+					optional := vf != nil && *vf.Optional
+					// Try to grab the referenced config map.
 					cm, err := s.resourceManager.GetConfigMap(vf.Name, pod.Namespace)
-					if vf.Optional != nil && !*vf.Optional && errors.IsNotFound(err) {
-						return fmt.Errorf("ConfigMap %s is required by Pod %s and does not exist", vf.Name, pod.Name)
-					}
-
 					if err != nil {
-						return fmt.Errorf("Error retrieving ConfigMap %s required by Pod %s: %s", vf.Name, pod.Name, err)
+						// We couldn't fetch the config map.
+						// However, if the key reference is optional we should not fail.
+						if optional {
+							// TODO: Emit an event (the contents of which possibly depend on whether the error is "NotFound", for clarity).
+							continue
+						}
+						// At this point we know the key reference is mandatory.
+						// Hence, we should return a meaningful error.
+						if errors.IsNotFound(err) {
+							return fmt.Errorf("required configmap %q not found", vf.Name)
+						}
+						return fmt.Errorf("failed to fetch required configmap %q: %v", vf.Name, err)
 					}
-
-					var ok bool
-					if c.Env[i].Value, ok = cm.Data[vf.Key]; !ok {
-						return fmt.Errorf("ConfigMap %s key %s is required by Pod %s and does not exist", vf.Name, vf.Key, pod.Name)
+					// At this point we have successfully fetched the target config map.
+					// We must now try to grab the requested key.
+					var (
+						keyExists bool
+						keyValue  string
+					)
+					if keyValue, keyExists = cm.Data[vf.Key]; !keyExists {
+						// The requested key does not exist.
+						// However, we should not fail if the key reference is optional.
+						if optional {
+							// TODO: Emit an event.
+							continue
+						}
+						// At this point we know the key reference is mandatory.
+						// Hence, we should fail.
+						return fmt.Errorf("configmap %q doesn't contain the %q key required by pod %s", vf.Name, vf.Key, pod.Name)
 					}
+					// Populate the environment variable and continue on to the next reference.
+					c.Env[i].Value = keyValue
 					continue
 				}
-
 				// Populate Secrets to Env
 				if e.ValueFrom.SecretKeyRef != nil {
 					vf := e.ValueFrom.SecretKeyRef
-					sec, err := s.resourceManager.GetSecret(vf.Name, pod.Namespace)
-					if vf.Optional != nil && !*vf.Optional && errors.IsNotFound(err) {
-						return fmt.Errorf("Secret %s is required by Pod %s and does not exist", vf.Name, pod.Name)
+					// Check whether the key reference is optional.
+					// This will control whether we fail when unable to read the requested key.
+					optional := vf != nil && *vf.Optional
+					// Try to grab the referenced secret.
+					cm, err := s.resourceManager.GetSecret(vf.Name, pod.Namespace)
+					if err != nil {
+						// We couldn't fetch the secret.
+						// However, if the key reference is optional we should not fail.
+						if optional {
+							// TODO: Emit an event (the contents of which possibly depend on whether the error is "NotFound", for clarity).
+							continue
+						}
+						// At this point we know the key reference is mandatory.
+						// Hence, we should return a meaningful error.
+						if errors.IsNotFound(err) {
+							return fmt.Errorf("required secret %q not found", vf.Name)
+						}
+						return fmt.Errorf("failed to fetch required secret %q: %v", vf.Name, err)
 					}
-					v, ok := sec.Data[vf.Key]
-					if !ok {
-						return fmt.Errorf("Secret %s key %s is required by Pod %s and does not exist", vf.Name, vf.Key, pod.Name)
+					// At this point we have successfully fetched the target secret.
+					// We must now try to grab the requested key.
+					var (
+						keyExists bool
+						keyValue  []byte
+					)
+					if keyValue, keyExists = cm.Data[vf.Key]; !keyExists {
+						// The requested key does not exist.
+						// However, we should not fail if the key reference is optional.
+						if optional {
+							// TODO: Emit an event.
+							continue
+						}
+						// At this point we know the key reference is mandatory.
+						// Hence, we should fail.
+						return fmt.Errorf("secret %q doesn't contain the %q key required by pod %s", vf.Name, vf.Key, pod.Name)
 					}
-					c.Env[i].Value = string(v)
+					// Populate the environment variable and continue on to the next reference.
+					c.Env[i].Value = string(keyValue)
 					continue
 				}
 

--- a/vkubelet/podcontroller.go
+++ b/vkubelet/podcontroller.go
@@ -195,7 +195,7 @@ func (pc *PodController) processNextWorkItem(ctx context.Context, workerId strin
 		if err := pc.syncHandler(ctx, key); err != nil {
 			if pc.workqueue.NumRequeues(key) < maxRetries {
 				// Put the item back on the work queue to handle any transient errors.
-				log.G(ctx).Warnf("requeuing %q due to failed sync", key)
+				log.G(ctx).Warnf("requeuing %q due to failed sync: %v", key, err)
 				pc.workqueue.AddRateLimited(key)
 				return nil
 			}

